pkg/model/imagereference.proto

```diff
@@ -0,0 +1,38 @@
+// Copyright 2021 The PipeCD Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax = "proto3";
+
+package pipe.model;
+option go_package = "github.com/pipe-cd/pipe/pkg/model";
+
+import "validate/validate.proto";
+
+message ImageReference {
+    // The generated unique identifier.
+    string id = 1 [(validate.rules).string.min_len = 1];
+    // The repository name.
+    string repo_name = 2 [(validate.rules).string.min_len = 1];
+    // The image digest.
+    string digest = 3;
```

Comment:
just a comment: This must be unique, so I initially thought it's better to use this as an id, while it's enough to be an optional field. Never mind.

Reply:
I found that the digest must not be unique. Unlike other general container registries, we have to handle multiple repositories (including gcr.io, ecr.amazonaws.com and docker hub). So it often happens that multiple references with the same digest to be registered.

```
$ docker images
REPOSITORY                                                         TAG                 IMAGE ID            CREATED             SIZE
nakabonne-test                                                     foo                 87c41dfd667d        13 days ago         194MB
000000000000.dkr.ecr.ap-northeast-1.amazonaws.com/nakabonne-test   foo                 87c41dfd667d        13 days ago         194MB
```
Really never mind!

Reply:
as far as I know, the combination of repository and tag is used to identify container images, the digest just used to compare 2 images, which have the same digest means they are the same. Image digest is hard to read/use so they do use repository and tag combination instead on refer to an image. In our case, digest should be just saved as external metadata to be used for future purposes.

Reply:
> Why the digest is optional?

Because we don't actually need this value, and if it's required then the user will have to get that value to set.
Our user does not only use the docker they may use bazel and other tools to build their container images.

> Why not use the digest as the ID?

An image digest is a hash of an image. So It depends on the image's manifest & layer.

```
docker images --digests
REPOSITORY                            TAG                   DIGEST                                                                    IMAGE ID       CREATED         SIZE
alpine                                latest                sha256:3c7497bf0c7af93428242d6176e8f7905f2201d8fc5861f45be7a346b5f23436   389fef711851   3 weeks ago     5.58MB
```

The same image (same digest) can be stored in multiple registries.